### ISSUE TITLE
stage/@type should allow multiple values

### ISSUE THREAD

**Issue by martindholmes** [TEIC/TEI#1395]
`<stage>` / @type is data.enumerated, with a semi val list. However, a single stage direction may comprise more than one category:

"Enter the King and all his Traine before the Gates."

This is "entrance" and "setting".

So the datatype should be modified to allow multiple values.

Also, `<stage>` should be a member of att.typed so that it also has @subtype.

**Comment by jamescummings** [comment]
1) If you are doing such detailed classifications shouldn't this be done with `@ana` pointing to a `<category>`?
2) Agreed. Anything with type, if possible should have att.typed and so subtype.

**Comment by PFSchaffner** [comment]
I agree with Martin here. His example does not constitute 'detailed classification' (suitable for feature structures) but rather the basic classification into types as recommended by the guidelines. Or so I think.

**Comment by jamescummings** [comment]
Isn't then he asking for a major change to `@type` everywhere it appears in att.typed?  I'm much more in love with URI pointers than magic words where feasible. ;-)

**Comment by martindholmes** [comment]
I'm not asking for anything major. The idea is that @type is inherited from att.typed by `<stage>`, which then modifies it to allow multiple values. No change to anything else.

**Comment by ebeshero** [comment]
The examples are helpful: yes, the stage element really ought to be able to hold multiple values for @type. 

And there other elements like stage that ought to belong to att.typed? Is there any reason that they don't?

Elisa

**Comment by jamescummings** [comment]
Maybe I'm misunderstanding. But if we give it att.typed and then modify `@type` locally to allow multiple values doesn't that mean that this then is significantly different to other type attributes?  I'm in favor of local modification for descs and valLists  and such. But if it is true here why is it not true for all instances of type?

**Comment by PFSchaffner** [comment]
There is, however, the slippery slope (James will not say): if a stage can rightly represent two different types of stage, is not the same likely true of (say) div, if not more so?

**Comment by PFSchaffner** [comment]
OH sorry. James in fact *did* say!

**Comment by jamescummings** [comment]
To be clear, I'm not saying we shouldn't open this can of worms, I'm just trying to read the label before we do so.  (And github is certainly easier to comment on from my phone on the busite home from work than sourceforge was.)

**Comment by martindholmes** [comment]
"Slippery slope" is not an argument against anything; we've had that discussion before. In fact, James made this very point himself at the FtF. If @type on `<stage>` is modified to have a different content model from @type anywhere else, it doesn't affect @type anywhere else. There may be similar arguments to be made for other instances of @type, but I'm not making them, and if nobody else does then there's nothing to worry about. If arguments are made for other contexts they should be addressed on their merits.

**Comment by jamescummings** [comment]
Is there any other instance of `@type` having multiple values? I'm not making the slippery slope argument really just pointing out this is a significant departure from current practice.

James


--
Dr James Cummings, Academic IT, University of Oxford

-----Original Message-----

**Comment by martindholmes** [comment]
I don't know if there are any other instances. Is there a principled objection that things can somehow only be of one type? 

It's possible that @type is not actually the right name for this classifying attribute on `<stage>`, and that it should be something else, but I can't think of anything suitable. @actions?

**Comment by hcayless** [comment]
My main concern with this is how @subtype interacts with @type when it has multiple values...

**Comment by PFSchaffner** [comment]
well, in typing divs we effectively use multiple values by simply combining them. E.g. we use div types "poem" and "preface" as common div types; a verse preface is so typed in our P3 (@type="verse preface" or @type="poem preface") which convert to P5 @type="verse_preface" or @type="prefatory_poem".

**Comment by martindholmes** [comment]
That's the strategy I've taken refuge in too, but it's horrible and arbitrary -- poem_preface or preface_poem? 

This arises out of converting some horrible old SGML to TEI.

**Comment by ebeshero** [comment]
If P3 permitted @type="poem preface", that seems better than the merging of the two words. I'd rather tokenize on a space than an underscore if I were processing it, not that it's much different, but perhaps we should keep distinct categories distinct.

As for distinguishing @type and @subtype, if there are multiple values permitted on each, does it make sense to recommend creating a meaningful array in which type values 1 2 3 are aligned with subtype values 1 2 3? Mind you, I'm not looking st what we already say in the guidelines about this--just thinking about meaningful and tidy typing and subtyping.

Elisa, council newbie

**Comment by hcayless** [comment]
Yeah, but then what happens if a @type has multiple @subtypes? I fear that way lies only madness…I’m not sure it makes sense to allow @subtype if a @type can have multiple values. Why wouldn’t @ana do the job for complex categorization like this?

That said, I’m not particularly against multi-types...

**Comment by martindholmes** [comment]
I'm increasingly convinced this is the wrong attribute name for this purpose. @action? @direction? @instruction? If we created a custom attribute that allowed multiple values, we could make the element a member of att.typed anyway; no existing files would be invalidated, but we could put a bit more work into creating a better semi list for the new attribute, and we wouldn't be violating the norms for @type. All in favour? Preferences for the attribute name?

**Comment by lb42** [comment]
Precisely so. If you want to say that this is a stage direction of two 
types A and B, you must either say
-- this is a new type A_B
-- this is type A, subtype A+B
-- this is type B, subtype B+A

The attribute is intended to simplify things, not complexify them. If 
you want to analyse their function in more detail, this is not an 
appropriate attribute to use unless you are willing to complexify your 
list of potential values.

And the "thin end of the wedge" argument does (I think)  apply. Another, 
less prejudicial, way of stating it would be to point to the number of 
other occasions on which we've tried to enforce a degree of consistency, 
so that similarly-named attributes always have similar behaviours. So  
making all @type attributes multi-valued, in the same way as all @target 
attributes are, might be a plausible proposal (except that I think 
Hugh's comment below makes it problematic)  ; making arbitrarily one of 
them multivalued is not.

Your fall back position of a totally different (and new) attribute 
@function or @actions or @effect or whatever we call it is much less 
implausible, but would leave us with two different ways of doing the 
same thing, unless you are also proposing to go theough the business of 
deprecating @type on stage as well.

So on balance, I think the least worst choice is @type="A_B". If you are 
really worried about the difficulty of parsing that, you could allow 
both @type="B_A" and @type="A_B" ....

**Comment by PFSchaffner** [comment]
I think that I agree with everyone.

I agree with Martin that multiple types of stage can
apply to a single instruction, even when you are
using @type to create a very simple and basic
taxonomy of stage directions. I.e., his use case is
not one of complex analysis, simply one of non-exclusive
categories.

I agree with Martin and James that there is no obvious
reason why <stage> should not join att.typed and get
@subtype.

I agree with Hugh that allowing multiple values in @typed
(whether here or on any element in att.typed) will play
havoc with the relationship between @type and @subtype.

And I agree with Lou (as usual) that the options are as
he has given them.

The only addition I would make is that <stage> suffers
other kinds of problematic overlap, notably with <speaker>
(many directions are both that and speaker designations,
with a large fuzzy area that different projects treat differently).
We (to the dismay of many) take a narrow view of <speaker>
and often split <stage> directions in order to carve offf a
bit of it into a <speaker>, when possible. The same approach
is *possible* here, even if ugly:

<stage type="entrance">Enter the King and all his Traine</stage>
<stage type="setting">before the Gates.</stage>

Or <stage> could be allowed to nest and segment itself with
subordinate <stage>s.

Or you could use <seg> within <stage> to isolate portions
with more specific functions:

<stage type="composite">
<seg type="entrance">Enter the King and all his Traine</seg>
<seg type="setting">before the Gates.</seg></stage>

All of these options seem preferable to changing the
data type of @type.

**Comment by martindholmes** [comment]
It's always a bit sad when the arguments for symmetry and coherence of the infrastructure (all @types should behave similarly, even in widely different use-cases) and inconvenience (change involves work, deprecation, etc.) trump a reasonable use-case. In this instance, I'm converting something from SGML, and one of the project principles has a long-standing antipathy to XML and TEI; it doesn't help to have to say that, in this case, what you can do perfectly straightforwardly in your SGML has to be done in a roundabout, indirect manner (@ana etc.) in TEI. I'd also add that arbitrarily demoting one of the values to @subtype is not a solution; one is not a subtype of the other. But I'll give up on this one. Closing the ticket.

**Comment by ebeshero** [comment]
I'm curious about usages of  `@ana` pointing to an external categories list, and whether that, too, introduces difficulties (such as in human readability of the TEI markup). I've seen multiple cases of things that belong to two major categories, and I can't imagine people applying @type and @subtype attributes in a way that wasn't attuned to project-specific needs--that they could readily process. The kinds of tangles we're imagining in the multiplicity of values for @type and @subtype seem more theoretical and less likely for project designers to want to apply in Real Life.

**Comment by hcayless** [comment]
I’m not sure I’d be that hasty. I don’t really see why `@type` on `<stage>` shouldn’t be special and allow multiple values. My only worry is that multi-valued `@type` isn’t really compatible with `@subtype`. I really don’t like the composite `@type="foo_bar"` solution. That’s horrible. If you have to do that, something’s clearly wrong.

**Comment by ebeshero** [comment]
(Can we reopen this?)

**Comment by hcayless** [comment]
Yeah, you’ve put your finger on the difficulty. We have trouble managing how our solution scales as complexity increases:

Simple problem—need to classify an element		Simple solution—@type
More complex—element classes aren’t so simple		Slightly more complex solution—@type & @subtype
Still more complex—multiple types					Completely different solution—@ana & category elements

This is kind of horrible, particularly if the encoder has a mixture of simple and complex element classification needs. On the other hand, scaling by imposing new syntax and semantics of @type and @subtype isn’t necessarily a good idea either. Maybe only having @type and not @subtype would be a better way to go (offering this as a for-instance, not as a real proposal, lest you all get nervous). Instead of

<div n="1" subtype="fragment" type="textpart">

we could have 

<div n="1" type="textpart fragment">

and it would be up to you to handle the semantics of the @types. That would break a lot of existing stuff though. It’s complicated. Welcome to the Council. We’re glad you’re here :-).

**Comment by lb42** [comment]
If Martin's users want to see an attribute in TEI XML which behaves in the same way as they are used to seeing in some ancient SGML system, and if they are not users to whom one can say "you are bonkers: go away", then clearly the only acceptable answer is to define a new attribute which has exactly the behaviour expected. Mapping an old SGML mukltivalued attribute to a new XML single-valued one was just wrong in the first place. But it's by no stretch of the imagination an argument for corrupting and confusing the well-established and perfectly reasonable way that @type/@subtype are currently defined.

**Comment by martindholmes** [comment]
Re-opening on the basis that we perhaps ought to discuss the creation of a special attribute. `@type` on `<stage>` is already a "special attribute"; it's not inherited from att.typed. It's arguable that the semantics of `@type` don't actually apply here anyway; we're not typing the stage direction, rather categorizing the various instructions it provides. 

@lb42, I do occasionally like to tell people they're bonkers, but in this case I don't think it applies. :-)

**Comment by ebeshero** [comment]
May I just ask why `@type` is a special attribute on `<stage>` in the first place? Why shouldn't it be the same `@type` inherited from att.typed? 

A related procedural question: is there a way for Council members (especially newbies) to look up the history of past Council deliberations on, say, the attributes to go on `<stage>`, or the development of elements for performance texts?

Elisa

**Comment by martindholmes** [comment]
The ability to inherit an attribute from a class and then override (change) it is relatively new. We have been slowly migrating the existing `@type` attributes which were declared directly on elements over to the system where they inherit from att.typed and then override, but it takes time; this one hasn't been looked at yet.

To look up the history of changes, there are four main resources:

1. The Council mailing list archives, which are public:
http://lists.tei-c.org/pipermail/tei-council/
2. The Council meeting minutes:
http://www.tei-c.org/Activities/Council/Meetings/index.xml
3. The commit history of the git repo (which includes the previous history from svn)
4. The tickets on GitHub (which also include everything from SourceForge)

**Comment by ebeshero** [comment]
Thanks for the context and the links, Martin! From what we've been discussing, it doesn't sound as if migrating the stage element to att.typed would resolve the problem, would it? (Or would it?) 
Elisa

**Comment by martindholmes** [comment]
Adding it to att.typed would give it `@subtype`, but as discussed, that won't really solve the problem. It would also mean that the argument that this particular `@type` is different from other `@type`s, and should therefore have a different content model, would be harder to sustain. But I think that argument is lost. I think it's inevitable that it will be added to att.typed; the only thing left to decide is whether it deserves a more specialized attribute with a different name that allows multiple values, or whether the need for that is not significant enough to warrant the resulting availability of more than method for doing the same thing.

**Comment by lb42** [comment]
(Incidentally, no-one seems to have noticed that the current &lt;valList> for stage/@type includes a value for precisely this case : "mixed". We're left with deciding how to indicate the components of tjhe mixture -- either as a @subtype or as some special new attribute yet to be invented)

**Comment by martindholmes** [comment]
I did know about 'mixed', but as you say, it doesn't solve the problem. Using `@type` for the the type when there's only one of them, but then leaping to a different attribute when there's more than one, is just mad.

**Comment by lb42** [comment]
It's not completely mad. You might consider "more than one of the above" to be a useful categorization without going into further detail.

**Comment by martindholmes** [comment]
I wouldn't, actually. If you list three values, I can tell by counting them that there's more than one. If you tell me "mixed", I only know there's more than one, but I don't know what they are; there's not much I can usefully do with that at all. I appreciate that this argument will inevitably lead to Hugh's suggestion that @type in att.typed should permit multiple values; perhaps that should be another ticket.

**Comment by lb42** [comment]
I didn't mean "you" Martin, I meant "some encoder". But I think the easiest way to resolve this is to recognise that the @type on &lt;stage> is not the same @type as we have in att.typed specifically in that it permits repetition. This seems less awful than making the att.typed/@type repeatable (which would introduce ambiguity in cases where @subtype is also used) and less awful than introducing a differently named local attribute (which would create a tsunami of compatibility issues). A differently named attribute is however the right solution long term: so we should eventually move to deprecating local @type in favour of it. 
If you agree, I've made the required changes in my local copy, and can check them in for next release!

**Comment by martindholmes** [comment]
That works for me, and it'll cover my needs for the project which raised the issue. Thanks!

**Comment by lb42** [comment]
Done. Let's hope we remember why we did this when we start revising attribute class memberships again.

**Comment by hcayless** [comment]
Would it be horrible if there were two classes containing a `@type`? One with `@type` and `@subtype` and the other with multi-valued `@type`. These would be mutually exclusive.

**Comment by lb42** [comment]
imho, yes it would be horrible. I would much rather have a locally-defined multi-valued attribute such as @function or @role or something with exactly the right semantics. In general, as I think of it, @type should be a single value. If a thing can have several @type values, then it's not a @type.

**Comment by hcayless** [comment]
As a programmer, the idea of something having multiple types seems like the
most natural thing in the world to me :-)

**Comment by lb42** [comment]
You're evidently not the same generation of programmer as me then!
